cmd/snap-update-ns/utils.go

```diff
@@ -362,6 +362,93 @@ func planWritableMimic(dir string) ([]*Change, error) {
 	return changes, nil
 }
 
+// execWritableMimic executes the plan for a writable mimic.
+// The result is a transformed mount namespace and a set of fake mount changes
+// that only exist in order to undo the plan.
+//
+// Certain assumptions are made about the plan, it must closely resemble that
+// created by planWritableMimic, in particular the sequence must look like this:
+//
+// - bind a directory aside into safekeeping location
+// - cover the original with tmpfs
+// - bind mount something from safekeeping location to an empty file or
+//   directory in the tmpfs; this step can repeat any number of times
+// - unbind the safekeeping location
+//
+// Apart from merely executing the plan a fake plan is returned for undo. The
+// undo plan skips the following elements as compared to the original plan:
+//
+// - the initial bind mount that constructs the safekeeping directory is gone
+// - the final unmount that removes the safekeeping directory
+// - the source of each of the bind mounts that re-populate tmpfs.
+//
+// In the event of a failure the undo plan is executed and an error is
+// returned. If the undo plan fails the function panics as it cannot fix the
+// system from an inconsistent state.
+func execWritableMimic(plan []*Change) ([]*Change, error) {
+	undoChanges := make([]*Change, 0, len(plan)-2)
+	for i, change := range plan {
+		if _, err := changePerform(change); err != nil {
+			recoveryUndoChanges := make([]*Change, 0, len(undoChanges)+1)
+			if i > 0 {
+				// The undo plan doesn't contain the entry for the initial bind
+				// mount of the safe keeping directory but we have already
+				// performed it. For this recovery phase we need to insert that
+				// in front of the undo plan manually.
+				recoveryUndoChanges = append(recoveryUndoChanges, plan[0])
+			}
+			recoveryUndoChanges = append(recoveryUndoChanges, undoChanges...)
+
+			// Drat, we failed! Let's undo everything according to our own undo
+			// plan, by following it in reverse order.
+			for j := len(recoveryUndoChanges) - 1; j >= 0; j-- {
+				recoveryUndoChange := recoveryUndoChanges[j]
+				// All the changes mount something, we need to reverse that.
+				// The "undo plan" is "a plan that can be undone" not "the plan
+				// for how to undo" so we need to flip the actions.
+				recoveryUndoChange.Action = Unmount
+				if _, err2 := changePerform(recoveryUndoChange); err2 != nil {
+					// Drat, we failed when trying to recover from an error.
+					// We cannot do anything at this stage.
+					panic(fmt.Errorf("cannot undo change %q while recovering from earlier error %v: %v", recoveryUndoChange, err, err2))
```

Comment:
I'm not convinced that panicking here is the right thing to do. Perhaps passing an error further up the stack would be a better solution. You should have more context there and can produce something more informative than this error and a stack trace.

Since it's a 'fatal-needing-a-panic error' this might warrant a separate type so that it's distinguishable. 

```
type FatalError struct {
    error
}

func IsFatalError(err error) bool {
    _, ok := err.(*FatalError)
    return ok
}

func NewFatalError(err error) *FatalError {
    return &FatalError{error: err}
}
```

working example: https://play.golang.org/p/nMTpTDMbHl

Reply:
Well, the caller up the stack will just error (which is what this will do). I can make the transformation but really, at this stage we cannot do anything useful.

Reply:
Done